pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -707,118 +707,145 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
-			return &status, nil
+			return status, nil
 		}
 
 		maxUnavailable, err = intstrutil.GetValueFromIntOrPercent(intstrutil.ValueOrDefault(set.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable, intstrutil.FromInt(1)), int(*set.Spec.Replicas), false)
 		if err != nil {
-			return &status, err
+			return status, err
 		}
 		// maxUnavailable should not be less than 1
 		if maxUnavailable < 1 {
 			maxUnavailable = 1
 		}
 	}
 
-	// refresh states for all pods
-	var modified bool
-	for _, pod := range pods {
-		refreshed, duration, err := ssc.refreshPodState(set, pod, updateRevision.Name)
-		if err != nil {
-			return &status, err
-		} else if duration > 0 {
-			durationStore.Push(getStatefulSetKey(set), duration)
+	minWaitTime := appsv1beta1.MaxMinReadySeconds * time.Second
+	unavailablePods := sets.NewString()
+	opts := &inplaceupdate.UpdateOptions{}
+	opts = inplaceupdate.SetOptionsDefaults(opts)
+	// counts any targets in the replicas that are unhealthy (terminated / in-place update not-ready / not running and
+	// ready for minReadySeconds) for checking if the count satisfied the MaxUnavailable limit.
+	for target := range replicas {
+		if replicas[target] == nil {
+			continue
 		}
-		if refreshed {
-			modified = true
+		if !isHealthy(replicas[target]) {
+			// 1. count pod as unavailable if it's unhealthy or terminating
+			unavailablePods.Insert(replicas[target].Name)
+		} else if completedErr := opts.CheckPodUpdateCompleted(replicas[target]); completedErr != nil {
+			// 2. count pod as unavailable if it's in-place updating and not ready
+			klog.V(4).Infof("StatefulSet %s/%s check Pod %s in-place update not-ready: %v",
+				set.Namespace,
+				set.Name,
+				replicas[target].Name,
+				completedErr)
+			unavailablePods.Insert(replicas[target].Name)
+		} else if isAvailable, waitTime := isRunningAndAvailable(replicas[target], minReadySeconds); !isAvailable {
+			// 3. count pod as unavailable if it's not available yet given the minReadySeconds requirement
+			unavailablePods.Insert(replicas[target].Name)
+			// make sure that we will wait for the first pod to get available
+			if waitTime != 0 && waitTime <= minWaitTime {
+				minWaitTime = waitTime
+				durationStore.Push(getStatefulSetKey(set), waitTime)
+			}
 		}
 	}
-	if modified {
-		return &status, nil
-	}
 
-	var unavailablePods []string
 	updateIndexes := sortPodsToUpdate(set.Spec.UpdateStrategy.RollingUpdate, updateRevision.Name, *set.Spec.Replicas, replicas)
 	klog.V(3).Infof("Prepare to update pods indexes %v for StatefulSet %s", updateIndexes, getStatefulSetKey(set))
-	minWaitTime := appsv1beta1.MaxMinReadySeconds * time.Second
 	// update pods in sequence
 	for _, target := range updateIndexes {
 
+		// the unavailable pods count exceed the maxUnavailable and the target is available, so we can't process it,
+		// wait for unhealthy Pods on update
+		if len(unavailablePods) >= maxUnavailable && !unavailablePods.Has(replicas[target].Name) {
+			klog.V(4).Infof(
```

Comment:
1. if "getPodRevision(replicas[target]) == updateRevision.Name", we can continue the loop. 
2. we can count the pods blocked by maxUnavailable, and log the error in one place so as to avoid duplicated error logging for every blocked pods

Reply:
[d4cecdd18060f7bcc3eeb97364ea3d77b93782d7..46e4e4111c32ab60a122f3e0fcdc5e9a74628bdd](https://github.com/openkruise/kruise/compare/d4cecdd18060f7bcc3eeb97364ea3d77b93782d7..46e4e4111c32ab60a122f3e0fcdc5e9a74628bdd)

1. changed
2. this log only be printed once per reconcilation when the first pod is blocked. By combining the first blocked pod with the L808 update indexes log, we can get all blocked pods if needed.